lib/Runtime/Library/JavascriptPromise.cpp

```diff
@@ -559,6 +559,147 @@ namespace Js
         return CreateThenPromise(promise, fulfillmentHandler, rejectionHandler, scriptContext);
     }
 
+    // Promise.prototype.finally as described in the draft of ES 2018 Section 25.4.5.3
+    Var JavascriptPromise::EntryFinally(RecyclableObject* function, CallInfo callInfo, ...)
+    {
+        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        ARGUMENTS(args, callInfo);
+        Assert(!(callInfo.Flags & CallFlags_New));
+
+        ScriptContext* scriptContext = function->GetScriptContext();
+
+        AUTO_TAG_NATIVE_LIBRARY_ENTRY(function, callInfo, _u("Promise.prototype.finally"));
+        //1. Let promise be the this value
+        //2. If Type(promise) is not Object, throw a TypeError exception
+        if (args.Info.Count < 1 || !JavascriptOperators::IsObject(args[0]))
+        {
+            JavascriptError::ThrowTypeError(scriptContext, JSERR_This_NeedObject, _u("Promise.prototype.finally"));
+        }
+
+        JavascriptLibrary* library = scriptContext->GetLibrary();
+        RecyclableObject* promise = RecyclableObject::UnsafeFromVar(args[0]);
+        //3. Let C be ? SpeciesConstructor(promise, %Promise%).
+        RecyclableObject* constructor = JavascriptOperators::SpeciesConstructor(promise, scriptContext->GetLibrary()->GetPromiseConstructor(), scriptContext);
+        //4. Assert IsConstructor(C)
+        Assert(JavascriptOperators::IsConstructor(constructor));
+
+        //5. If IsCallable(onFinally) is false
+        //a. Let thenFinally be onFinally
+        //b. let catchFinally be onFinally 
+        //6. Else,
+        //  a. Let thenFinally be a new built-in function object as defined in ThenFinally Function.
+        //  b. Let catchFinally be a new built-in function object as defined in CatchFinally Function.
+        //  c. Set thenFinally and catchFinally's [[Constructor]] internal slots to C.
+        //  d. Set thenFinally and catchFinally's [[OnFinally]] internal slots to onFinally.
+
+        Var thenFinally;
+        Var catchFinally;
```

Comment:
I'd like if these were initialized to `nullptr` and have an assert after the conditional that each is not `nullptr` anymore. Should cause no overhead in release build.

Reply:
fixed